File: app/src/main/java/edu/upc/mcia/publications/data/repository/PublicationRepository.java
package edu.upc.mcia.publications.data.repository;

import java.util.List;

import edu.upc.mcia.publications.data.model.Author;
import edu.upc.mcia.publications.data.model.Publication;
import edu.upc.mcia.publications.data.remote.ApiManager;
import edu.upc.mcia.publications.data.remote.dto.Page;
import edu.upc.mcia.publications.data.remote.dto.PublicationDto;
import io.reactivex.Observable;

public class PublicationRepository {

    private static PublicationRepository INSTANCE = null;

    private final AuthorRepository mAuthorRepository;
    private final PublisherRepository mPublisherRepository;

    private PublicationRepository() {
        mAuthorRepository = AuthorRepository.getInstance();
        mPublisherRepository = PublisherRepository.getInstance();
    }

    public static PublicationRepository getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new PublicationRepository();
        }
        return INSTANCE;
    }

    public Observable<Page<Publication>> search(int pageNumber, String query) {
        return ApiManager.getRemoteApi().getPublications(pageNumber, query)
                .flatMap(this::fillPublication);
    }

    private Observable<Page<Publication>> fillPublication(Page<PublicationDto> pageDto) {
        return Observable
                .fromIterable(pageDto.getContent())
                .concatMap(dto ->
                        Observable.zip(
                                getAuthors(dto.getAuthorIds()),
                                mPublisherRepository.findById(dto.getPublisherId()),
                                (authors, publisher) -> Publication.from(dto, authors, publisher)))

                .toList()
                .toObservable()
                .map(list -> new Page<>(list, pageDto.getMetadata()));
    }

    private Observable<List<Author>> getAuthors(List<String> ids) {
        return Observable
                .fromIterable(ids)
                .concatMap(mAuthorRepository::findById)
                .toList().toObservable();
    }

}

File: app/src/main/java/edu/upc/mcia/publications/data/repository/PublicationRepository.kt
package edu.upc.mcia.publications.data.repository

import edu.upc.mcia.publications.data.model.Author
import edu.upc.mcia.publications.data.model.Publication
import edu.upc.mcia.publications.data.model.Publisher
import edu.upc.mcia.publications.data.remote.ApiManager
import edu.upc.mcia.publications.data.remote.dto.Page
import edu.upc.mcia.publications.data.remote.dto.PublicationDto
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import java.util.*

object PublicationRepository {

    private val mAuthorRepository: AuthorRepository
    private val mPublisherRepository: PublisherRepository

    private val mData = HashMap<String, Publication>()

    init {
        mAuthorRepository = AuthorRepository
        mPublisherRepository = PublisherRepository.getInstance()
    }

    fun search(pageNumber: Int, query: String): Observable<Page<Publication>> {
        return ApiManager.remoteApi.getPublications(pageNumber, query)
                .flatMap({ this.fillPublicationPage(it) })
                .doOnNext { mData += it.content.map { it.id to it } }
    }

    fun findById(publicationId: String): Observable<Publication> {
        mData[publicationId]?.let { return Observable.just(it) } ?:
                return Observable.just(publicationId)
                        .flatMap { ApiManager.remoteApi.getPublication(it) }
                        .flatMap { fillPublication(it) }
    }

    private fun fillPublication(dto: PublicationDto): Observable<Publication> {
        return Observable.zip<List<Author>, Publisher, Publication>(
                getAuthors(dto.authorIds),
                mPublisherRepository.findById(dto.publisherId),
                BiFunction { authors, publisher -> Publication.from(dto, authors, publisher) })

    }

    private fun fillPublicationPage(pageDto: Page<PublicationDto>): Observable<Page<Publication>> {
        return Observable
                .fromIterable(pageDto.content)
                .concatMap(this::fillPublication)
                .toList()
                .toObservable()
                .map { list -> Page(list, pageDto.metadata) }
    }

    private fun getAuthors(ids: List<String>): Observable<List<Author>> {
        return Observable
                .fromIterable(ids)
                .concatMap<Author>({ mAuthorRepository.findById(it) })
                .toList().toObservable()
    }

}
